tests/integration/utils/network.py

```diff
@@ -29,41 +30,33 @@ def next_range(self, number_of_ports: int) -> "PortRangeInclusive":
 
         return PortRangeInclusive(next_start, next_end)
 
+    @contextmanager
+    def allocate_port(self) -> int:
+        """Find a random port in the range `PortRangeInclusive`.
 
-# To find a good port range use the following:
-#
-#   curl --silent 'https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.txt' | \
-#       egrep -i -e '^\s*[0-9]+-[0-9]+\s*unassigned' | \
-#       awk '{print $1}'
-#
-KAFKA_PORT_RANGE = PortRangeInclusive(48700, 48800)
-ZK_PORT_RANGE = KAFKA_PORT_RANGE.next_range(100)
-REGISTRY_PORT_RANGE = ZK_PORT_RANGE.next_range(100)
-TESTS_PORT_RANGE = REGISTRY_PORT_RANGE.next_range(100)
-
-
-def get_random_port(*, port_range: PortRangeInclusive, blacklist: List[int]) -> int:
-    """Find a random port in the range `PortRangeInclusive`.
-
-    Note:
-        This function is *not* aware of the ports currently open in the system,
-        the blacklist only prevents two services of the same type to randomly
-        get the same ports for *a single test run*.
-
-        Because of that, the port range should be chosen such that there is no
-        system service in the range. Also note that running two sessions of the
-        tests with the same range is not supported and will lead to flakiness.
-    """
-    assert port_range.start <= port_range.end, f"{port_range.start} must be less-than-or-equal to {port_range.end}"
-
-    # +1 because randint is inclusive for both ends
-    ports_in_range = (port_range.end - port_range.start) + 1
-    assert len(blacklist) < ports_in_range, f"no free ports available. Range {port_range}, blacklist: {blacklist}"
-
-    value = random.randint(port_range.start, port_range.end)
-    while value in blacklist:
-        value = random.randint(port_range.start, port_range.end)
-    return value
+        Note:
+            This function is *not* aware of the ports currently open in the system,
+            the blacklist only prevents two services of the same type to randomly
+            get the same ports for *a single test run*.
+
+            Because of that, the port range should be chosen such that there is no
+            system service in the range. Also note that running two sessions of the
+            tests with the same range is not supported and will lead to flakiness.
+        """
+        # +1 because randint is inclusive for both ends
+        ports_in_range = (self.end - self.start) + 1
+
+        if ports_in_range <= len(self.allocated):
+            raise RuntimeError(f"no free ports available. Range {self}, allocated: {self.allocated}")
+
+        port = random.randint(self.start, self.end)
+        while port in self.allocated:
+            port = random.randint(self.start, self.end)
+        self.allocated.add(port)
+
+        yield port
+
+        self.allocated.remove(port)
```

Comment:
I suppose in this case this is the best approach, if an exception is thrown on the yield something went wrong, and we can't be sure the port was properly released